legion_core/src/world.rs

```diff
@@ -903,6 +1178,122 @@ impl Default for World {
     fn default() -> Self { Self::new() }
 }
 
+/// Describes how to handle a clone_from. Allows the user to transform components from one type
+/// to another and provide their own implementation for cloning/transforming
+pub trait CloneImpl {
+    /// When a component of the provided `component_type` is encountered, we will transfer data
+    /// from it into the returned component type. For a basic clone implementation, this function
+    /// should return the same type as was passed into it
+    fn map_component_type(
+        &self,
+        component_type_id: ComponentTypeId,
+    ) -> (ComponentTypeId, ComponentMeta);
+
+    /// When called, the implementation should copy the data from src_data to dst_data. The
+    /// src_world and src_entities are provided so that other components on the same Entity can
+    /// be looked up. The dst_resources are provided so that any required side effects to resources
+    /// (like registering a physics body into a physics engine) can be implemented.
+    #[allow(clippy::too_many_arguments)]
+    fn clone_components(
+        &self,
+        src_world: &World,
+        src_component_storage: &ComponentStorage,
+        src_component_storage_indexes: core::ops::Range<ComponentIndex>,
+        src_type: ComponentTypeId,
+        src_entities: &[Entity],
+        dst_entities: &[Entity],
+        src_data: *const u8,
+        dst_data: *mut u8,
+        num_components: usize,
+    );
+}
+
+/// Used along with CloneImpl, allows receiving results from a clone_from or clone_from_single call
+pub trait CloneImplResult {
+    /// For every entity that is copied, this function will be called, passing the entity in the
+    /// source and destination worlds
+    fn add_result(&mut self, src_entity: Entity, dst_entity: Entity);
+}
+
+/// Used along with CloneImpl, allows specifying that certain entities in the receiving world should
+/// be replaced with entities from the source world
+pub trait EntityReplacePolicy<'s> {
+    /// This function must return all entities in the source world that will replace data in the
+    /// destination world
+    fn src_entities<'a>(&'s self) -> Box<dyn Iterator<Item = Entity> + 'a>
+    where
+        's: 'a;
+
+    /// This function must return all entities in the destination world that will be replaced
+    fn dst_entities<'a>(&'s self) -> Box<dyn Iterator<Item = Entity> + 'a>
+    where
+        's: 'a;
+
+    /// When called, if the provided source entity is intended to replace an entity in the
```

Comment:
This sentence reads a bit difficult. Maybe remove 'when called' part, because that is obvious and allows us to reduce the commas.